### PR TITLE
fix #4680 : Custom key deserialiser registered for `Object.class` is ignored on nested JSON

### LINKED ISSUE #4680: Custom key deserialiser registered for `String.class` is ignored when Map key type not defined
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

I'm working on a piece of software which accepts arbitrary JSON, flattens it, and then writes it to CSV. The main issue here is that we don't know the schema of the JSON—it can be anything. This, in turn, results in untyped deserialization.

Before starting processing of the parsed object, the field names must be sanitised (yes, it's crucial to sanitise the field names during the parsing phase). Unfortunately `UntypedObjectDeserializerNR` ignores the custom key deserialiser I'm trying to provide:

```java
package libraries;

import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.KeyDeserializer;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Map;

public class ObjectMapperTest {

    @Test
    void customKeyDeserializerShouldBeUsedWhenTypeNotDefined() throws Exception {
        // given
        var json = """
                {
                    "name*": "Erik",
                    "address*": {
                        "city*": {
                            "id*": 1,
                            "name*": "Berlin"
                        },
                        "street*": "Elvirastr"
                    }
                }
                                
                """;

        var objectMapper = new ObjectMapper();
        var keySanitizationModule = new SimpleModule("key-sanitization");
        keySanitizationModule.addKeyDeserializer(String.class, new KeyDeserializer() {
            @Override
            public Object deserializeKey(String key, DeserializationContext ctxt) {
                return key.replace("*", "_");
            }
        });
        objectMapper.registerModule(keySanitizationModule);

        // when
        var result = objectMapper.readValue(json, Object.class);

        // then
        Assertions.assertEquals("Erik", ((Map<String, Object>) result).get("name_"));

        var addressMap = (Map<String, Object>) ((Map<String, Object>) result).get("address_");
        Assertions.assertEquals("Elvirastr", addressMap.get("street_"));

        var cityMap = (Map<String, Object>) addressMap.get("city_");
        Assertions.assertEquals(1, cityMap.get("id_"));
        Assertions.assertEquals("Berlin", cityMap.get("name_"));
    }
}
```

I have 2 questions in regards of this behaviour: 
1. Is that something expected?
2. Is there a workaround for the problem in question?

### Version Information

2.14.3

### Reproduction

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JooHyukKim**
fixes #4680

Still have below three tests failing.

```log
[ERROR] Errors: 
[ERROR]   DeepNestingUntypedDeserTest.testFormerlyTooDeepUntypedWithArray » StackOverflow
[ERROR]   DeepNestingUntypedDeserTest.testFormerlyTooDeepUntypedWithObject » StackOverflow
[ERROR]   TestAbstractTypes.testDefaultingRecursive:137 » JsonMapping (was java.lang.NullPointerException) (through reference chain: java.util.TreeMap["x"])
```

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java [hunk 1]
```diff
@@ -191,18 +218,32 @@ public JsonDeserializer<?> createContextual(DeserializationContext ctxt,
...
-            return new UntypedObjectDeserializer(this, preventMerge);
+            untyped = new UntypedObjectDeserializer(this, preventMerge);
+        }
+        if (keyDeser != null) {
+            if (untyped == null) {
+                untyped = new UntypedObjectDeserializer(this, keyDeser);
+            } else {
+                untyped = new UntypedObjectDeserializer(untyped, keyDeser);
+            }
+        }
```
Comment (JooHyukKim) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:243]: @cowtowncoder This is what I came up with atm. I was hoping maybe you might think of better approach than this? 😅

First, null/non-null check for `keyDeser` is done at line 228 and line 237
Second, declaring local variable `untyped` seems a bit unnatural.

**Comment by cowtowncoder** [comment]
Ok, we might be able to make this work. But one quick idea: how about adding failing unit test first, in a separate PR? I could check that in first. And then hopefully help figure out what to do for actual problem.

FWTW, I suspect this will need to go in 2.19. But we can start PR against 2.18.

File: src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java [hunk 2]
```diff
@@ -191,18 +218,32 @@ public JsonDeserializer<?> createContextual(DeserializationContext ctxt,
         // 14-Jun-2017, tatu: [databind#1625]: may want to block merging, for root value
         boolean preventMerge = (property == null)
                 && Boolean.FALSE.equals(ctxt.getConfig().getDefaultMergeable(Object.class));
+        // 31-Aug-2024: [databind#4680] Allow custom key deserializer for Object.class deserialization
+        KeyDeserializer keyDeser = ctxt.findKeyDeserializer(SimpleType.constructUnsafe(Object.class), property);
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:222]: I think this needs to see if we get the "default" key deserializer -- I forget how exactly it is done, maybe regular `MapDeserializer` has an example -- and if so, ignore it (re-set to `null`).

**Comment by JooHyukKim** [comment]
> how about adding failing unit test first, in a separate PR?

Makes sense! Made #4685 against 2.18 (so that we can verify behavior since 2.18) @cowtowncoder

File: src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java (hunk 2 above)
Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:222]: Thanks! Added [commit](https://github.com/FasterXML/jackson-databind/pull/4684/commits/a1fe9cc888a4bffb64a604b500a7cc7577b02173) according to your suggestio and we are down to one last case 👌🏼👌🏼

```log
[ERROR] Errors: 
[ERROR]   UntypedDeserializationTest.testUntypedWithCustomScalarDesers:263 » NullPointer
```

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:222]: Ah, simple mistake, didn't check if we had _customKeyDeserializer not null

**Comment by JooHyukKim** [comment]
Marked as ready for review and waiting for 2.19 branch out :))

File: src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java [hunk 3]
```diff
@@ -508,6 +548,8 @@ protected Object mapObject(JsonParser p, DeserializationContext ctxt) throws IOE
             LinkedHashMap<String, Object> result = new LinkedHashMap<>(2);
             result.put(key1, value1);
             return result;
+        } else if (_customKeyDeserializer != null) {
+            key2 = (String) _customKeyDeserializer.deserializeKey(key2, ctxt);
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:552]: @JooHyukKim  I think same check needed for 2 cases further down too?
(and test should check that first 5 keys per level get custom deserialized as well)

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java:552]: Yes, seems like it. Will do

Comment (JooHyukKim) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/UntypedObjectDeserializer.java]: Applied ur review and did minor plummeting (with a new helper function) via https://github.com/FasterXML/jackson-databind/pull/4684/commits/1f7dfbc4dfa08a6b936e03e866f3ecfe6fc8c58c. Thanks!

**Comment by cowtowncoder** [comment]
Looks decent, need to re-base against 2.19 now.

**Comment by JooHyukKim** [comment]
Rebase done! 👌🏼

**Comment by JooHyukKim** [comment]
Cleaned up and improved some comments.
Thank you for the tweaks always, @cowtowncoder !

**Comment by cowtowncoder** [comment]
Hoping to get back to this issue once we resolve the pile of issues for 2.18.1. Not forgotten. :)

**Comment by cowtowncoder** [comment]
@JooHyukKim minor merge conflicts to resolve. Hoping to get back to reviewing this.

**Comment by JooHyukKim** [comment]
Resolved conflicts, thanks!

**Review by cowtowncoder — APPROVED** [review]
LGTM!

**Comment by cowtowncoder** [comment]
I think this is ready -- will merge!

**Pull request merged**